### PR TITLE
Init VSCode Extension

### PR DESCRIPTION
**Pull request by tbezman**
This PR aims to scaffold out an MVP for the Relay LSP VSCode Extension.

## High Level Changes
- Added new `vscode-extension` in the root dir containing all the extension code.
- We're using typescript in this new directory. Spoke this over with @captbaritone and it seems like we're cool with it.
  - Main benefit is we get all the types published by the VSCode team for free without having to wait for published flow defs.
- Added a basic LSP Client in `extension.ts` which looks for a relay binary and starts it up. Things seem to be working well in the Coinbase codebase.
- Added a `launch.json` so contributors can easily test the extension with `F5`. It also lets you see the debug output from the LSP Server directly in the opened window. It's really sweet.

## Follow Ups
I left a bunch of TODOs in the source if you want some more insight.

- [ ] Use VSCode config instead of environment variables from the launch json
- [ ] Support VSCode workspaces (don't use `workspace.rootPath`). Not exactly sure how to handle this but will be looking at the flow extension for more guidance.
- [ ] Better error message handling
- [ ] Status Bar support
- [x] Fix go to definition on fields / types
- [ ] Add command to manually restart the LSP Client.
- [x] Add a Github Action to do type checking / build check / prettier / linting

### CODE REVIEW

File: vscode-extension/.vscode/launch.json
Comment (captbaritone) [inline vscode-extension/.vscode/launch.json:12]: I'm guessing we should omit this?

File: vscode-extension/package.json
Comment (captbaritone) [inline vscode-extension/package.json:3]: Maybe make this "0.0.1" for now.

File: vscode-extension/package.json
Comment (captbaritone) [inline vscode-extension/package.json:23]: Do we need this? People can just enable/disable the extension I think? Is there some other value?

File: vscode-extension/package.json
Comment (captbaritone) [inline vscode-extension/package.json:28]:
```suggestion
          "description": "Controls what is logged to the Output Channel."
```

Maybe we can use the enum/enum description as outlined here? https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:35]: Love this.

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:55]: 👍

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:73]: In that case, would they also not have a relay binary? Maybe we could assume that no Relay binary means not a Relay project, but if they have a Relay binary, we expect to find a config?

**edit** Ah, you pointed that out below.

File: vscode-extension/src/utils.ts
Comment (captbaritone) [inline vscode-extension/src/utils.ts:23]: Maybe add a comment that this mirrors what we're doing in [`relay-compiler/index.js`](https://github.com/facebook/relay/blob/14550c8bd80e53369852db66d7681f13d6dd5f84/packages/relay-compiler/index.js)? If we make a change in one place we probably want to make it in both.

Maybe also add a comment in that file pointing back here?

File: vscode-extension/src/utils.ts
Comment (captbaritone) [inline vscode-extension/src/utils.ts:47]: I like that this will correctly handle the case relay is installed at the root of a Yarn workspace!

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:82]:
```suggestion
            "Relay LSP client connection got closed unexpectedly.",

```

File: vscode-extension/src/utils.ts
Comment (captbaritone) [inline vscode-extension/src/utils.ts:56]: Is this copied from some extension? If not, it might be worthing seeing if we can find someone to test this on Windows just to double check. I assume it should work, but an infinite loop is also a bad error to have.

File: vscode-extension/.vscode/launch.json
Comment (captbaritone) [inline vscode-extension/.vscode/launch.json:5]: Is this string shown anywhere? Maybe this should be "Relay Extension"?

**Review by captbaritone — CHANGES_REQUESTED** [review]
Overall this looks quite reasonable! First round of feedback left. I think addressing these would be enough to merge and then iterate.

File: vscode-extension/package.json
Comment (captbaritone) [inline vscode-extension/package.json:2]: Is this the name that gets used in the VSCode Marketplace? I suspect this should be human readable. Maybe "Relay GraphQL"? Not sure what the best name is here. Might need to think about it.

File: vscode-extension/src/extension.ts
Comment (captbaritone) [reply vscode-extension/src/extension.ts:82]: Oh, another thing we should explicitly handle is what happens if you are on an older version of Relay. When this extension comes out, it will only work with the very newest version of Relay. Ideally we could check if they are on an old version and give them a message letting them know they need to upgrade. Otherwise they will probably get a very confusing error where the binary just exits with an error code.

(We can tackle this in a followup PR)

File: vscode-extension/.vscode/launch.json
Comment (captbaritone) [reply vscode-extension/.vscode/launch.json:12]: Confirmed, this causes the extension to crash (unless you know to edit it) but removing it allowed it to work!
Screen Shot 2022-04-04 at 7 49 18 PM

**Comment by captbaritone** [comment]
Got it working on a little toy side project!

https://user-images.githubusercontent.com/162735/161670634-a193a85f-28f3-43fc-8a68-7d10d116cbf9.mov

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:43]: Can we try adding `markdown: {isTrusted: true}`? I'm noticing that hover items are not correctly rendering the content as markdown
Screen Shot 2022-04-04 at 9 01 14 PM
.

File: vscode-extension/.vscode/launch.json
Comment (tbezman) [reply vscode-extension/.vscode/launch.json:12]: this is temporary and probably helped you run it locally? probably fine to keep it until we get proper vscode config setup.

File: vscode-extension/package.json
Comment (tbezman) [reply vscode-extension/package.json:23]: yeah vscode can disable extensions per workspace now
image

File: vscode-extension/.vscode/launch.json
Comment (captbaritone) [reply vscode-extension/.vscode/launch.json:12]: Once I removed it my server started correctly. I don't think we need it.

Comment (captbaritone) [reply vscode-extension/.vscode/launch.json:12]: Oh, I see what you're saying. This lets you use a locally compiled version of the binary. I have ` "relay-compiler": "0.0.0-main-7a27722d",` in my package.json, so it's using a recent enough version that the compiler supports the LSP.

Maybe leave it but comment it out? (I think VSCode's JSON parser supports comments?)

File: vscode-extension/src/utils.ts
Comment (tbezman) [reply vscode-extension/src/utils.ts:56]: yeah I'm definitely making an assumption here. i'll try to run some tests on my gaming pc tonight

File: vscode-extension/package.json
Comment (tbezman) [reply vscode-extension/package.json:2]: So the flow one has some hyphenated name https://github.com/flow/flow-for-vscode/blob/main/package.json

I think they're like bundle identifiers in iOS. I'm fine with anything. I'll defer to you 
image

File: vscode-extension/src/utils.ts
Comment (captbaritone) [reply vscode-extension/src/utils.ts:59]: If you have any question about it, you could add a counter and increment each loop and bail if it reaches something absurd like 5,000.

File: vscode-extension/src/extension.ts
Comment (tbezman) [reply vscode-extension/src/extension.ts:43]: okay I think that did the trick. does this look right? 
image

Comment (captbaritone) [reply vscode-extension/src/extension.ts:43]: Perfect

File: vscode-extension/package.json
Comment (captbaritone) [reply vscode-extension/package.json:2]: Oh interesting. I wonder where the human readable bits come from.

Comment (orta) [reply vscode-extension/package.json:2]: Use `displayName` in the `package.json` - https://code.visualstudio.com/api/references/extension-manifest

File: vscode-extension/src/utils.ts
Comment (tbezman) [reply vscode-extension/src/utils.ts:59]: good call

File: vscode-extension/package.json
Comment (tbezman) [reply vscode-extension/package.json:2]: Thanks @orta !

File: vscode-extension/src/utils.ts
Comment (captbaritone) [inline vscode-extension/src/utils.ts:49]: ❤️

File: vscode-extension/src/utils.ts
Comment (tbezman) [reply vscode-extension/src/utils.ts:66]: image

File: vscode-extension/package.json
Comment (captbaritone) [inline vscode-extension/package.json:3]: I think the fact that it's in the VSCode Marketplace is enough to communicate that it's for VSCode. I'll chat with the team about what we want here. I'm thinking "Relay GraphQL". Fine to merge this as is and update once we decide.

File: vscode-extension/src/utils.ts
Comment (tbezman) [reply vscode-extension/src/utils.ts:66]: looks like it works on windows. but i'll keep the counter in as a fail safe just in case.

File: vscode-extension/package.json
Comment (alunyov) [inline vscode-extension/package.json:2]: can we just use `relay` here?

File: vscode-extension/package.json
Comment (alunyov) [inline vscode-extension/package.json:36]: do we have any plans integration typecheker with Github actions?

Also we probably need to add lint/prettier?

Comment (tbezman) [reply vscode-extension/package.json:36]: yeah i'll add those as followups

Comment (tbezman) [reply vscode-extension/package.json:36]: @alunyov any preference on including these changes in this PR vs a follow up?

Comment (alunyov) [reply vscode-extension/package.json:36]: Follow up is good! Thank you!

Comment (tbezman) [reply vscode-extension/package.json]: @alunyov I went ahead and added the following to this PR

- prettier config which is basically the same as the flow one (just had to change the parser)
- eslint config from airbnb with some slight modifications
- typecheck

All this is passing in CI

Comment (tbezman) [reply vscode-extension/package.json]: rough timing, totally cool if you want me to pull these commits out and add them in a follow up instead.

Comment (alunyov) [reply vscode-extension/package.json]: Oh, this is great! Thanks, it's fine to keep it here

File: vscode-extension/src/extension.ts
Comment (captbaritone) [inline vscode-extension/src/extension.ts:1]: Can you add this header?

```
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
 ```

File: vscode-extension/src/utils.ts
Comment (captbaritone) [inline vscode-extension/src/utils.ts:1]: Same header as above:

```
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
 ```

File: vscode-extension/.eslintrc
Comment (captbaritone) [inline vscode-extension/.eslintrc:12]: Can wee add https://eslint.org/docs/rules/linebreak-style set to "unix"?

Maybe it's already fixed by Prettier, but when I first imported this PR I got an error from an internal linter:

Screen Shot 2022-04-05 at 9 28 25 PM

File: .github/workflows/ci.yml
Comment (alunyov) [inline .github/workflows/ci.yml:11]: I think we can just have a single job here:
- vscode-extension

File: .github/workflows/ci.yml
Comment (alunyov) [inline .github/workflows/ci.yml:60]: and then we won't need this

File: vscode-extension/src/extension.ts
Comment (alunyov) [inline vscode-extension/src/extension.ts:26]: same with the log-level? Should we use the value from the `configuration`?

File: vscode-extension/.eslintrc
Comment (tbezman) [reply vscode-extension/.eslintrc:12]: yeah I think this should be fixed with the latest & greatest

File: vscode-extension/src/extension.ts
Comment (tbezman) [reply vscode-extension/src/extension.ts:26]: def in the follow up PR

**Comment by tbezman** [comment]
@alunyov I've consolidated the CI jobs into one. lemme know if that looks alright.

**Review by alunyov — APPROVED** [review]
Thank you @tbezman!

**Pull request closed without merging**
